In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
base = pd.read_csv("churn.csv", index_col='RowNumber')
base

CustomerId    Surname  CreditScore Geography  Gender  Age  Tenure  \
RowNumber                                                                      
1            15634602   Hargrave          619    France  Female   42       2   
2            15647311       Hill          608     Spain  Female   41       1   
3            15619304       Onio          502    France  Female   42       8   
4            15701354       Boni          699    France  Female   39       1   
5            15737888   Mitchell          850     Spain  Female   43       2   
...               ...        ...          ...       ...     ...  ...     ...   
9996         15606229   Obijiaku          771    France    Male   39       5   
9997         15569892  Johnstone          516    France    Male   35      10   
9998         15584532        Liu          709    France  Female   36       7   
9999         15682355  Sabbatini          772   Germany    Male   42       3   
10000        15628319     Walker          792    France  Female   28       4   

             Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber                                                        
1               0.00              1          1               1   
2           83807.86              1          0               1   
3          159660.80              3          1               0   
4               0.00              2          0               0   
5          125510.82              1          1               1   
...              ...            ...        ...             ...   
9996            0.00              2          1               0   
9997        57369.61              1          1               1   
9998            0.00              1          0               1   
9999        75075.31              2          1               0   
10000      130142.79              1          1               0   

           EstimatedSalary  Exited  
RowNumber                           
1                101348.88       1  
2                112542.58       0  
3                113931.57       1  
4                 93826.63       0  
5                 79084.10       0  
...                    ...     ...  
9996              96270.64       0  
9997             101699.77       0  
9998              42085.58       1  
9999              92888.52       1  
10000             38190.78       0  

[10000 rows x 13 columns]

In [3]:
arr_delete = ['CustomerId', 'Surname']

base.drop(arr_delete, axis=1, inplace = True)
base

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
1                  619    France  Female   42       2       0.00   
2                  608     Spain  Female   41       1   83807.86   
3                  502    France  Female   42       8  159660.80   
4                  699    France  Female   39       1       0.00   
5                  850     Spain  Female   43       2  125510.82   
...                ...       ...     ...  ...     ...        ...   
9996               771    France    Male   39       5       0.00   
9997               516    France    Male   35      10   57369.61   
9998               709    France  Female   36       7       0.00   
9999               772   Germany    Male   42       3   75075.31   
10000              792    France  Female   28       4  130142.79   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
1                      1          1               1        101348.88       1  
2                      1          0               1        112542.58       0  
3                      3          1               0        113931.57       1  
4                      2          0               0         93826.63       0  
5                      1          1               1         79084.10       0  
...                  ...        ...             ...              ...     ...  
9996                   2          1               0         96270.64       0  
9997                   1          1               1        101699.77       0  
9998                   1          0               1         42085.58       1  
9999                   2          1               0         92888.52       1  
10000                  1          1               0         38190.78       0  

[10000 rows x 11 columns]

In [4]:
#tipos de dados
base.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 1 to 10000
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Geography        10000 non-null  object 
 2   Gender           10000 non-null  object 
 3   Age              10000 non-null  int64  
 4   Tenure           10000 non-null  int64  
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int64  
dtypes: float64(2), int64(7), object(2)
memory usage: 937.5+ KB


In [5]:
#estatísticas
base.describe().round(2)

CreditScore       Age    Tenure    Balance  NumOfProducts  HasCrCard  \
count     10000.00  10000.00  10000.00   10000.00       10000.00   10000.00   
mean        650.53     38.92      5.01   76485.89           1.53       0.71   
std          96.65     10.49      2.89   62397.41           0.58       0.46   
min         350.00     18.00      0.00       0.00           1.00       0.00   
25%         584.00     32.00      3.00       0.00           1.00       0.00   
50%         652.00     37.00      5.00   97198.54           1.00       1.00   
75%         718.00     44.00      7.00  127644.24           2.00       1.00   
max         850.00     92.00     10.00  250898.09           4.00       1.00   

       IsActiveMember  EstimatedSalary   Exited  
count        10000.00         10000.00  10000.0  
mean             0.52        100090.24      0.2  
std              0.50         57510.49      0.4  
min              0.00            11.58      0.0  
25%              0.00         51002.11      0.0  
50%              1.00        100193.92      0.0  
75%              1.00        149388.25      0.0  
max              1.00        199992.48      1.0

## Analisando apenas clientes que saíram do banco

In [6]:
base[base["Exited"] == 1].describe().round(0)

CreditScore     Age  Tenure   Balance  NumOfProducts  HasCrCard  \
count       2037.0  2037.0  2037.0    2037.0         2037.0     2037.0   
mean         645.0    45.0     5.0   91109.0            1.0        1.0   
std          100.0    10.0     3.0   58361.0            1.0        0.0   
min          350.0    18.0     0.0       0.0            1.0        0.0   
25%          578.0    38.0     2.0   38340.0            1.0        0.0   
50%          646.0    45.0     5.0  109349.0            1.0        1.0   
75%          716.0    51.0     8.0  131433.0            2.0        1.0   
max          850.0    84.0    10.0  250898.0            4.0        1.0   

       IsActiveMember  EstimatedSalary  Exited  
count          2037.0           2037.0  2037.0  
mean              0.0         101466.0     1.0  
std               0.0          57912.0     0.0  
min               0.0             12.0     1.0  
25%               0.0          51908.0     1.0  
50%               0.0         102461.0     1.0  
75%               1.0         152423.0     1.0  
max               1.0         199808.0     1.0

Da base geral, de 10.000 clientes, apenas 2.037 clientes saíram do banco. 
Credit Score = a pontuação máxima de crédito encontrada no geral do banco é de 850, também é a dos clientes que sairam do banco. 


## Correlação de Colunas e Clientes que saíram: 

### CreditScore 

In [7]:
base[(base["CreditScore"] >= 425) & (base["Exited"] ==1)]

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
1                  619    France  Female   42       2       0.00   
3                  502    France  Female   42       8  159660.80   
6                  645     Spain    Male   44       8  113755.78   
17                 653   Germany    Male   58       1  132602.88   
23                 510     Spain  Female   38       4       0.00   
...                ...       ...     ...  ...     ...        ...   
9982               498   Germany    Male   42       3  152039.70   
9983               655   Germany  Female   46       7  137145.12   
9992               597    France  Female   53       4   88381.21   
9998               709    France  Female   36       7       0.00   
9999               772   Germany    Male   42       3   75075.31   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
1                      1          1               1        101348.88       1  
3                      3          1               0        113931.57       1  
6                      2          1               0        149756.71       1  
17                     1          1               0          5097.67       1  
23                     1          1               0        118913.53       1  
...                  ...        ...             ...              ...     ...  
9982                   1          1               1         53445.17       1  
9983                   1          1               0        115146.40       1  
9992                   1          1               0         69384.71       1  
9998                   1          0               1         42085.58       1  
9999                   2          1               0         92888.52       1  

[2005 rows x 11 columns]

In [8]:
base[(base["CreditScore"] >= 645) & (base["Exited"] ==1)]

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
6                  645     Spain    Male   44       8  113755.78   
17                 653   Germany    Male   58       1  132602.88   
44                 834    France  Female   49       2  131394.56   
47                 829   Germany  Female   27       9  112045.67   
54                 655   Germany    Male   41       8  125561.97   
...                ...       ...     ...  ...     ...        ...   
9961               795   Germany  Female   33       9  104552.72   
9963               702     Spain    Male   44       9       0.00   
9983               655   Germany  Female   46       7  137145.12   
9998               709    France  Female   36       7       0.00   
9999               772   Germany    Male   42       3   75075.31   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
6                      2          1               0        149756.71       1  
17                     1          1               0          5097.67       1  
44                     1          0               0        194365.76       1  
47                     1          1               1        119708.21       1  
54                     1          0               0        164040.94       1  
...                  ...        ...             ...              ...     ...  
9961                   1          1               1        120853.83       1  
9963                   1          0               0         59207.41       1  
9983                   1          1               0        115146.40       1  
9998                   1          0               1         42085.58       1  
9999                   2          1               0         92888.52       1  

[1034 rows x 11 columns]

Levei em consideração a metade de 850, o valor de 425, a nível de comparação, pois é o valor máximo encontrado de crédito dos clientes gerais, não sendo possível utilizar 100(o que seria o óbvio, mas não fica claro diante os dados.)

Foi possível observar que 2.005, ou seja 98.42% dos clientes evasivos tinham o score acima ou igual a metade do valor.

Já analisando pela média encontrada pelo .info() foi encontrado 1.034 clientes evasivos, 50.76%.

### Age

In [9]:
base[base["Exited"] == 1].describe().round(0)

CreditScore     Age  Tenure   Balance  NumOfProducts  HasCrCard  \
count       2037.0  2037.0  2037.0    2037.0         2037.0     2037.0   
mean         645.0    45.0     5.0   91109.0            1.0        1.0   
std          100.0    10.0     3.0   58361.0            1.0        0.0   
min          350.0    18.0     0.0       0.0            1.0        0.0   
25%          578.0    38.0     2.0   38340.0            1.0        0.0   
50%          646.0    45.0     5.0  109349.0            1.0        1.0   
75%          716.0    51.0     8.0  131433.0            2.0        1.0   
max          850.0    84.0    10.0  250898.0            4.0        1.0   

       IsActiveMember  EstimatedSalary  Exited  
count          2037.0           2037.0  2037.0  
mean              0.0         101466.0     1.0  
std               0.0          57912.0     0.0  
min               0.0             12.0     1.0  
25%               0.0          51908.0     1.0  
50%               0.0         102461.0     1.0  
75%               1.0         152423.0     1.0  
max               1.0         199808.0     1.0

O perfil de clientes evasivos com base na idade, varia muito tambem. O mínimo tem 18 anos, como esperado para ter conta em banco, o máximo do perfil de idade desses usuários é de 84 anos, mas a maioria fica entre 38 à 51 anos.

### Tenure

In [10]:
base[(base["Tenure"] >= 5) & (base["Exited"] ==1) & (base["Gender"] == "Female") ]

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
3                  502    France  Female   42       8  159660.80   
42                 465    France  Female   51       8  122522.32   
47                 829   Germany  Female   27       9  112045.67   
48                 637   Germany  Female   39       9  137843.80   
91                 647     Spain  Female   44       5       0.00   
...                ...       ...     ...  ...     ...        ...   
9935               692   Germany  Female   43      10  118588.83   
9957               520    France  Female   46      10   85216.61   
9961               795   Germany  Female   33       9  104552.72   
9983               655   Germany  Female   46       7  137145.12   
9998               709    France  Female   36       7       0.00   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
3                      3          1               0        113931.57       1  
42                     1          0               0        181297.65       1  
47                     1          1               1        119708.21       1  
48                     1          1               1        117622.80       1  
91                     3          1               1        174205.22       1  
...                  ...        ...             ...              ...     ...  
9935                   1          1               1        161241.65       1  
9957                   1          1               0        117369.52       1  
9961                   1          1               1        120853.83       1  
9983                   1          1               0        115146.40       1  
9998                   1          0               1         42085.58       1  

[615 rows x 11 columns]

In [11]:
base[(base["Tenure"] >= 5) & (base["Exited"] ==1) & (base["Gender"] == "Female") ]

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
3                  502    France  Female   42       8  159660.80   
42                 465    France  Female   51       8  122522.32   
47                 829   Germany  Female   27       9  112045.67   
48                 637   Germany  Female   39       9  137843.80   
91                 647     Spain  Female   44       5       0.00   
...                ...       ...     ...  ...     ...        ...   
9935               692   Germany  Female   43      10  118588.83   
9957               520    France  Female   46      10   85216.61   
9961               795   Germany  Female   33       9  104552.72   
9983               655   Germany  Female   46       7  137145.12   
9998               709    France  Female   36       7       0.00   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
3                      3          1               0        113931.57       1  
42                     1          0               0        181297.65       1  
47                     1          1               1        119708.21       1  
48                     1          1               1        117622.80       1  
91                     3          1               1        174205.22       1  
...                  ...        ...             ...              ...     ...  
9935                   1          1               1        161241.65       1  
9957                   1          1               0        117369.52       1  
9961                   1          1               1        120853.83       1  
9983                   1          1               0        115146.40       1  
9998                   1          0               1         42085.58       1  

[615 rows x 11 columns]

O Tenure, marca um tipo de score de confiança do usuário, vai de 0 à 10. 
com a média sendo 5, a base possui 1.093 clientes evasivos com confiança maior ou igual a média, ditos, boa confiança, sendo 53.65%. 

ainda nesse perfil, de tenure "bom", 615 são mulheres, a maioria. 

In [12]:
base[base["Exited"] == 1].describe().round(0)

CreditScore     Age  Tenure   Balance  NumOfProducts  HasCrCard  \
count       2037.0  2037.0  2037.0    2037.0         2037.0     2037.0   
mean         645.0    45.0     5.0   91109.0            1.0        1.0   
std          100.0    10.0     3.0   58361.0            1.0        0.0   
min          350.0    18.0     0.0       0.0            1.0        0.0   
25%          578.0    38.0     2.0   38340.0            1.0        0.0   
50%          646.0    45.0     5.0  109349.0            1.0        1.0   
75%          716.0    51.0     8.0  131433.0            2.0        1.0   
max          850.0    84.0    10.0  250898.0            4.0        1.0   

       IsActiveMember  EstimatedSalary  Exited  
count          2037.0           2037.0  2037.0  
mean              0.0         101466.0     1.0  
std               0.0          57912.0     0.0  
min               0.0             12.0     1.0  
25%               0.0          51908.0     1.0  
50%               0.0         102461.0     1.0  
75%               1.0         152423.0     1.0  
max               1.0         199808.0     1.0

### HasCrCard

In [13]:
base[(base["HasCrCard"] == 1) & (base["Exited"] == 1 )]

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
1                  619    France  Female   42       2       0.00   
3                  502    France  Female   42       8  159660.80   
6                  645     Spain    Male   44       8  113755.78   
8                  376   Germany  Female   29       4  115046.74   
17                 653   Germany    Male   58       1  132602.88   
...                ...       ...     ...  ...     ...        ...   
9976               610   Germany    Male   50       1  113957.01   
9982               498   Germany    Male   42       3  152039.70   
9983               655   Germany  Female   46       7  137145.12   
9992               597    France  Female   53       4   88381.21   
9999               772   Germany    Male   42       3   75075.31   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
1                      1          1               1        101348.88       1  
3                      3          1               0        113931.57       1  
6                      2          1               0        149756.71       1  
8                      4          1               0        119346.88       1  
17                     1          1               0          5097.67       1  
...                  ...        ...             ...              ...     ...  
9976                   2          1               0        196526.55       1  
9982                   1          1               1         53445.17       1  
9983                   1          1               0        115146.40       1  
9992                   1          1               0         69384.71       1  
9999                   2          1               0         92888.52       1  

[1424 rows x 11 columns]

Clientes que saíram do banco e possuiam cartão de crédito foram: 1.424 numa porcentagem de 69.90%

### IsActiveMember

In [14]:
base[(base["IsActiveMember"] == 0) & (base["Exited"] == 1 )]

CreditScore Geography  Gender  Age  Tenure    Balance  \
RowNumber                                                          
3                  502    France  Female   42       8  159660.80   
6                  645     Spain    Male   44       8  113755.78   
8                  376   Germany  Female   29       4  115046.74   
17                 653   Germany    Male   58       1  132602.88   
23                 510     Spain  Female   38       4       0.00   
...                ...       ...     ...  ...     ...        ...   
9963               702     Spain    Male   44       9       0.00   
9976               610   Germany    Male   50       1  113957.01   
9983               655   Germany  Female   46       7  137145.12   
9992               597    France  Female   53       4   88381.21   
9999               772   Germany    Male   42       3   75075.31   

           NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
RowNumber                                                                     
3                      3          1               0        113931.57       1  
6                      2          1               0        149756.71       1  
8                      4          1               0        119346.88       1  
17                     1          1               0          5097.67       1  
23                     1          1               0        118913.53       1  
...                  ...        ...             ...              ...     ...  
9963                   1          0               0         59207.41       1  
9976                   2          1               0        196526.55       1  
9983                   1          1               0        115146.40       1  
9992                   1          1               0         69384.71       1  
9999                   2          1               0         92888.52       1  

[1302 rows x 11 columns]

1.302 membros (63.91%) que encerram a conta, já não eram membros ativos no banco. 

In [15]:
def categoria_quali_quanti(coluna):
    colunabase = base.groupby([coluna, "Exited"]).size()
    coluna = base.groupby(coluna).size()
    return (colunabase / coluna*100).reset_index()

### Gender

Mais mulheres saíram do banco, mas não são informações relevantes

### Geography

In [135]:
categoria_quali_quanti("Geography").round(2)

Geography  Exited      0
0    France       0  83.85
1    France       1  16.15
2   Germany       0  67.56
3   Germany       1  32.44
4     Spain       0  83.33
5     Spain       1  16.67

A maior evasão por localidade é por parte da Alemanha

### NumOfProducts

In [136]:
categoria_quali_quanti("NumOfProducts").round(2)

NumOfProducts  Exited       0
0              1       0   72.29
1              1       1   27.71
2              2       0   92.42
3              2       1    7.58
4              3       0   17.29
5              3       1   82.71
6              4       1  100.00

100% das pessoas que tem 4 produtos sairam do banco. 

### EstimatedSalary

Os salários são muito distintos. 